pkg/table-rule-selector/trie_selector.go

```diff
@@ -0,0 +1,458 @@
+			if n.question == nil {
+				return nil, errors.NotFoundf("pattern %v", pattern)
+			}
+
+			items = append(items, n.question)
+			n = n.question.child
+		default:
+			item, ok := n.characters[pattern[i]]
+			if !ok {
+				return nil, errors.NotFoundf("pattern %v", pattern)
+			}
+			items = append(items, item)
+			n = item.child
+		}
+	}
+
+	return items, nil
+}
+
+// AllRules implements Selector's AllRules
+func (t *trieSelector) AllRules() (map[string]interface{}, map[string]map[string]interface{}) {
+	var (
+		tableRules  = make(map[string]map[string]interface{})
+		schemaNodes = make(map[string]*node)
+		schemaRules = make(map[string]interface{})
+		word        []byte
+	)
+	t.RLock()
+	t.travel(t.root, word, schemaRules, schemaNodes)
+
+	for schema, n := range schemaNodes {
+		rules, ok := tableRules[schema]
+		if !ok {
+			rules = make(map[string]interface{})
+		}
+
+		word = word[:0]
+		t.travel(n, word, rules, nil)
+		if len(rules) > 0 {
+			tableRules[schema] = rules
+		}
+	}
+	t.RUnlock()
+	return schemaRules, tableRules
+}
+
+func (t *trieSelector) travel(n *node, word []byte, rules map[string]interface{}, nodes map[string]*node) {
+	if n == nil {
+		return
+	}
+
+	if n.asterisk != nil {
+		pattern := append(word, asterisk)
+		insertMatchedItemIntoMap(string(pattern), n.asterisk, rules, nodes)
+	}
+
+	if n.question != nil {
+		pattern := append(word, question)
+		insertMatchedItemIntoMap(string(pattern), n.question, rules, nodes)
+		t.travel(n.question.child, pattern, rules, nodes)
+	}
+
+	for char, item := range n.characters {
+		pattern := append(word, char)
+		if item != nil {
+			insertMatchedItemIntoMap(string(pattern), item, rules, nodes)
+			t.travel(item.child, pattern, rules, nodes)
+		}
+	}
+}
+
+func (t *trieSelector) matchNode(n *node, s string, mr *matchedResult) {
+	if n == nil {
+		return
+	}
+
+	var (
+		ok     bool
+		entity *item
+	)
+	for i := range s {
+		if n.asterisk != nil {
+			appendMatchedItem(n.asterisk, mr)
+		}
+
+		if n.question != nil {
+			if i == len(s)-1 {
+				appendMatchedItem(n.question, mr)
+			}
+
+			t.matchNode(n.question.child, s[i+1:], mr)
+		}
+
+		entity, ok = n.characters[s[i]]
+		if !ok {
+			return
+		}
+		n = entity.child
+	}
+
+	if entity != nil {
+		appendMatchedItem(entity, mr)
+	}
+
+	if n.asterisk != nil {
+		appendMatchedItem(n.asterisk, mr)
+	}
+}
+
+func appendMatchedItem(entity *item, mr *matchedResult) {
+	if entity.rule != nil {
+		mr.rules = append(mr.rules, entity.rule)
+	}
+
+	if entity.nextLevel != nil {
+		mr.nodes = append(mr.nodes, entity.nextLevel)
+	}
+}
+
+func insertMatchedItemIntoMap(pattern string, entity *item, rules map[string]interface{}, nodes map[string]*node) {
+	if rules != nil && entity.rule != nil {
+		rules[pattern] = entity.rule
+	}
+
+	if nodes != nil && entity.nextLevel != nil {
+		nodes[pattern] = entity.nextLevel
+	}
+}
+
+func (t *trieSelector) addToCache(key string, rules RuleSet) {
+	t.cache[key] = rules
+	if len(t.cache) > maxCacheNum {
```

Comment:
do you mean 
```
for literal := range t.cache {
    if len(t.cache) > maxCacheNum {
         delete(t.cache, literal)
    } else {
        break
    }
}
```

Reply:
and this may delete the key add to cache just now.

Reply:
only delete one cache, it's a poor cache implement, and I think it's rarely used